File: lib/data/models/alarm.dart
import 'dart:convert';
import 'dart:io';

import 'package:ff_alarm/globals.dart';
import 'package:ff_alarm/ui/utils/updater.dart';
import 'package:floor/floor.dart';
import 'package:flutter/material.dart';
import 'package:geolocator/geolocator.dart';

@entity
class Alarm {
  @primaryKey
  final String id;
  String get server => id.split(' ')[0];
  int get idNumber => int.parse(id.split(' ')[1]);

  String type;

  String word;

  DateTime date;

  int number;

  String address;
  Position? get positionFromAddressIfCoordinates {
    var split = address.split(',');
    if (split.length != 2) return null;
    try {
      double? lat = double.tryParse(split[0]);
      double? lon = double.tryParse(split[1]);
      if (lat == null || lon == null) return null;
      return Position(
        latitude: lat,
        longitude: lon,
        accuracy: 0,
        altitude: 0,
        altitudeAccuracy: 0,
        heading: 0,
        headingAccuracy: 0,
        speed: 0,
        speedAccuracy: 0,
        timestamp: DateTime.now(),
        floor: 0,
        isMocked: false,
      );
    } catch (_) {
      return null;
    }
  }

  List<String> notes;

  List<int> units;
  List<String> get unitProperIds => units.map((e) => "$server $e").toList();

  Map<int, AlarmResponse> responses;
  AlarmResponse? get ownResponse {
    AlarmResponse? response;
    for (var entry in responses.entries) {
      String personId = "$server ${entry.key}";
      if (Globals.localPersons.containsKey(personId)) {
        response = entry.value;
        break;
      }
    }
    return response;
  }

  int updated;

  bool get responseTimeExpired => date.isBefore(DateTime.now().subtract(const Duration(hours: 1)));

  Alarm({
    required this.id,
    required this.type,
    required this.word,
    required this.date,
    required this.number,
    required this.address,
    required this.notes,
    required this.units,
    required this.responses,
    required this.updated,
  });

  static const Map<String, String> jsonShorts = {
    "server": "s",
    "id": "i",
    "type": "t",
    "word": "w",
    "date": "d",
    "number": "n",
    "address": "a",
    "notes": "no",
    "units": "u",
    "responses": "r",
    "updated": "up",
  };

  factory Alarm.fromJson(Map<String, dynamic> json) {
    return Alarm(
      id: "${json[jsonShorts["server"]]} ${json[jsonShorts["id"]]}",
      type: json[jsonShorts["type"]],
      word: json[jsonShorts["word"]],
      date: DateTime.fromMillisecondsSinceEpoch(json[jsonShorts["date"]]),
      number: json[jsonShorts["number"]],
      address: json[jsonShorts["address"]],
      notes: List<String>.from(json[jsonShorts["notes"]]),
      units: List<int>.from(json[jsonShorts["units"]]),
      updated: json[jsonShorts["updated"]],
      responses: () {
        Map<int, AlarmResponse> result = {};
        Map<String, dynamic> decoded = json[jsonShorts["responses"]];
        decoded.forEach((key, value) {
          var response = AlarmResponse.fromJson(value);
          result[int.parse(key)] = response;
        });
        return result;
      }(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      jsonShorts["server"]!: server,
      jsonShorts["id"]!: idNumber,
      jsonShorts["type"]!: type,
      jsonShorts["word"]!: word,
      jsonShorts["date"]!: date.millisecondsSinceEpoch,
      jsonShorts["number"]!: number,
      jsonShorts["address"]!: address,
      jsonShorts["notes"]!: notes,
      jsonShorts["units"]!: units,
      jsonShorts["updated"]!: updated,
      jsonShorts["responses"]!: () {
        Map<String, dynamic> result = {};
        responses.forEach((key, value) {
          result[key.toString()] = value.toJson();
        });
        return result;
      }(),
    };
  }

  factory Alarm.inflateFromString(String input) {
    final decodeBase64Json = base64.decode(input);
    final decodedZipJson = gzip.decode(decodeBase64Json);
    final originalJson = utf8.decode(decodedZipJson);
    final Map<String, dynamic> json = jsonDecode(originalJson);
    return Alarm.fromJson(json);
  }

  Future<AlarmOption> getAlertOption(bool shouldNotify) async {
    DateTime now = DateTime.now();
    if (date.isAfter(now.subtract(const Duration(minutes: 15)))) {
      if (type.startsWith('Test')) {
        bool? muted = Globals.prefs.getBool('alarms_testsMuted');
        if (muted == true) return AlarmOption.silent;
      } else {
        bool? muted = Globals.prefs.getBool('alarms_muted');
        if (muted == true) return AlarmOption.silent;
      }

      return shouldNotify ? AlarmOption.alert : AlarmOption.silent;
    }

    if (date.isAfter(now.subtract(const Duration(hours: 24)))) return AlarmOption.silent;
    return AlarmOption.none;
  }

  static Future<List<Alarm>> getBatched({
    bool Function(Alarm)? filter,
    int? limit,
    String? startingId,
  }) async {
    var alarms = <Alarm>[];

    String lowestId = startingId ?? "\u{10FFFF}";
    const batchSize = 50;
    while (true) {
      var newAlarms = await Globals.db.alarmDao.getWithLowerIdThan(lowestId, batchSize);
      if (newAlarms.isEmpty) break;

      var toAdd = <Alarm>[];
      for (var alarm in newAlarms) {
        if (filter == null || filter(alarm)) toAdd.add(alarm);
        if (alarm.id.compareTo(lowestId) < 0) lowestId = alarm.id;

        if (limit != null && toAdd.length + alarms.length >= limit) break;
      }

      alarms.addAll(toAdd);
      if ((limit != null && alarms.length >= limit) || newAlarms.length < batchSize) break;
    }

    alarms.sort((a, b) => b.date.compareTo(a.date));

    return alarms;
  }

  static Future<List<Alarm>> getAll({bool Function(Alarm)? filter}) => getBatched(filter: filter);

  static Stream<List<Alarm>> getAllStreamed({int delay = 100}) async* {
    String lowestId = "\u{10FFFF}";
    const batchSize = 25;
    while (true) {
      var newAlarms = await Globals.db.alarmDao.getWithLowerIdThan(lowestId, batchSize);
      if (newAlarms.isEmpty) break;

      for (var alarm in newAlarms) {
        if (alarm.id.compareTo(lowestId) < 0) lowestId = alarm.id;
      }

      newAlarms.sort((a, b) => b.date.compareTo(a.date));
      yield newAlarms;
      if (newAlarms.length < batchSize) break;

      await Future.delayed(Duration(milliseconds: delay));
    }
  }

  static Future<void> update(Alarm alarm, bool bc) async {
    // guarantee, that:
    // id contains a space
    // id split 0 != null
    // id split 1 != 0
    var splits = alarm.id.split(' ');
    if (splits.length != 2 || splits[0].isEmpty || splits[1].isEmpty) return;
    if (splits[0] == 'null' || splits[1] == '0') return;
    if (int.tryParse(splits[1]) == null) return;

    var existing = await Globals.db.alarmDao.getById(alarm.id);

    if (existing != null) {
      if (existing.updated >= alarm.updated) return;
      await Globals.db.alarmDao.updates(alarm);
    } else {
      await Globals.db.alarmDao.inserts(alarm);
    }

    if (!bc) return;
    UpdateInfo(UpdateType.alarm, {alarm.id});
  }

  static Future<void> delete(String alarmId, bool bc) async {
    await Globals.db.alarmDao.deleteById(alarmId);

    if (!bc) return;
    UpdateInfo(UpdateType.alarm, {alarmId});
  }

  static Future<int?> getAmount(String server) => Globals.db.alarmDao.getAmountWithServer(server);
}

enum AlarmOption {
  alert,
  silent,
  none,
}

/// Each responder gives a single response to each alarm
class AlarmResponse {
  /// The note left to be visible for other responders of the station they are not "Not going" to
  /// If the response is "Not going" to all stations, the note is visible to all responders
  String note;

  /// The time at which the response was given
  DateTime time;

  /// The type of response given, for each station
  Map<int, AlarmResponseType> responses;

  AlarmResponse({
    required this.note,
    required this.time,
    required this.responses,
  });

  static const Map<String, String> jsonShorts = {
    "note": "n",
    "time": "t",
    "responses": "r",
  };

  factory AlarmResponse.fromJson(Map<String, dynamic> json) {
    return AlarmResponse(
      note: json[jsonShorts["note"]],
      time: DateTime.fromMillisecondsSinceEpoch(json[jsonShorts["time"]]),
      responses: () {
        Map<int, AlarmResponseType> responses = {};
        json[jsonShorts["responses"]].forEach((key, value) {
          responses[int.parse(key)] = AlarmResponseType.values[value];
        });
        return responses;
      }(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      jsonShorts["note"]!: note,
      jsonShorts["time"]!: time.millisecondsSinceEpoch,
      jsonShorts["responses"]!: responses.map((key, value) => MapEntry(key.toString(), value.index)),
    };
  }

  ({int? stationId, AlarmResponseType responseType}) getResponseInfo() {
    int? stationId;
    AlarmResponseType? responseType;

    bool anyNotSet = false;
    for (var response in responses.entries) {
      if (response.value != AlarmResponseType.notReady && response.value != AlarmResponseType.notSet) {
        responseType = response.value;
        stationId = response.key;
        break;
      }

      if (response.value == AlarmResponseType.notSet) anyNotSet = true;
    }

    if (anyNotSet) responseType = AlarmResponseType.notSet;
    responseType ??= AlarmResponseType.notReady;

    return (stationId: stationId, responseType: responseType);
  }
}

enum AlarmResponseType {
  onStation(0),
  under5(5),
  under10(10),
  under15(15),
  onCall(-1),
  notReady(-2),
  notSet(-3);

  final int timeAmount;

  const AlarmResponseType(this.timeAmount);

  Color get color {
    switch (this) {
      case AlarmResponseType.onStation:
        return Colors.green;
      case AlarmResponseType.under5:
        return Colors.cyanAccent;
      case AlarmResponseType.under10:
        return Colors.yellow;
      case AlarmResponseType.under15:
        return Colors.orange;
      case AlarmResponseType.onCall:
        return Colors.purpleAccent;
      case AlarmResponseType.notReady:
        return Colors.red;
      case AlarmResponseType.notSet:
        return Colors.grey;
    }
  }

  String get name {
    switch (this) {
      case AlarmResponseType.onStation:
        return 'An Wache';
      case AlarmResponseType.under5:
        return '< 5 Min';
      case AlarmResponseType.under10:
        return '< 10 Min';
      case AlarmResponseType.under15:
        return '< 15 Min';
      case AlarmResponseType.onCall:
        return 'Auf Abruf dazu';
      case AlarmResponseType.notReady:
        return 'Nicht bereit';
      case AlarmResponseType.notSet:
        return 'Nicht gesetzt';
    }
  }
}

File: ios/notifications/Alarm.swift
//
//  Alarm.swift
//  notifications
//
//  Created by Konstanius on 21.03.24.
//

import Foundation
import SQLite
import Gzip

class Alarm {
    var id: String
    var type: String
    var word: String
    var date: Date
    var number: Int
    var address: String
    var notes: [String]
    var units: [Int]
    var responses: [Int: AlarmResponse]
    var updated: Date

    init(id: String, type: String, word: String, date: Date, number: Int, address: String, notes: [String], units: [Int], responses: [Int: AlarmResponse], updated: Date) {
        self.id = id
        self.type = type
        self.word = word
        self.date = date
        self.number = number
        self.address = address
        self.notes = notes
        self.units = units
        self.responses = responses
        self.updated = updated
    }

    static let jsonShorts = [
        "server": "s",
        "id": "i",
        "type": "t",
        "word": "w",
        "date": "d",
        "number": "n",
        "address": "a",
        "notes": "no",
        "units": "u",
        "responses": "r",
        "updated": "up",
    ]

    init(json: [String: Any]) {
        self.id = json[Alarm.jsonShorts["server"]!] as! String + " " + String(json[Alarm.jsonShorts["id"]!] as! Int)
        self.type = json[Alarm.jsonShorts["type"]!] as! String
        self.word = json[Alarm.jsonShorts["word"]!] as! String
        self.date = Date(timeIntervalSince1970: TimeInterval(json[Alarm.jsonShorts["date"]!] as! Int))
        self.number = json[Alarm.jsonShorts["number"]!] as! Int
        self.address = json[Alarm.jsonShorts["address"]!] as! String
        self.notes = json[Alarm.jsonShorts["notes"]!] as! [String]
        self.units = json[Alarm.jsonShorts["units"]!] as! [Int]
        self.updated = Date(timeIntervalSince1970: TimeInterval(json[Alarm.jsonShorts["updated"]!] as! Int))
        self.responses = {
            var result = [Int: AlarmResponse]()
            let decoded = json[Alarm.jsonShorts["responses"]!] as! [String: Any]
            decoded.forEach { key, value in
                let response = AlarmResponse(json: value as! [String: Any])
                result[Int(key)!] = response
            }
            return result
        }()
    }

    func toJson() -> [String: Any] {
        let tempServer = id.split(separator: " ")[0]
        let tempId = Int(id.split(separator: " ")[1])!
        return [
            Alarm.jsonShorts["id"]!: tempId,
            Alarm.jsonShorts["server"]!: tempServer,
            Alarm.jsonShorts["type"]!: type,
            Alarm.jsonShorts["word"]!: word,
            Alarm.jsonShorts["date"]!: Int(date.timeIntervalSince1970),
            Alarm.jsonShorts["number"]!: number,
            Alarm.jsonShorts["address"]!: address,
            Alarm.jsonShorts["notes"]!: notes,
            Alarm.jsonShorts["units"]!: units,
            Alarm.jsonShorts["updated"]!: Int(updated.timeIntervalSince1970),
            Alarm.jsonShorts["responses"]!: {
                var result = [String: Any]()
                responses.forEach { key, value in
                    result[String(key)] = value.toJson()
                }
                return result
            }()
        ]
    }

    static func inflateFromString(input: String) -> Alarm {
        let decodeBase64Json = Data(base64Encoded: input)!
        let decodedZipJson = try! decodeBase64Json.gunzipped()
        let decodedUtf8 = String(data: decodedZipJson, encoding: .utf8)!
        let decodedJson = try! JSONSerialization.jsonObject(with: decodedUtf8.data(using: .utf8)!, options: []) as! [String: Any]
        let alarm = Alarm(json: decodedJson)
        return alarm
    }

    func getAlertOption(prefs: [String: Any], shouldNotify: Bool) -> AlarmOption {
        let now = Date()
        if date > now.addingTimeInterval(-15 * 60) {
            if word.starts(with: "Test") {
                if prefs["alarms_testsMuted"] as? Bool == true {
                    return .silent
                }
            } else {
                if prefs["alarms_muted"] as? Bool == true {
                    return .silent
                }
            }

            return shouldNotify ? .alert : .silent
        }
        if date > now.addingTimeInterval(-24 * 60 * 60) {
            return .silent
        }
        return .none
    }

    enum AlarmOption {
        case alert
        case silent
        case none
    }

    static func insert(alarm: Alarm) {
        let alarms = Table("Alarm")
        let id = Expression<String>("id")
        let type = Expression<String>("type")
        let word = Expression<String>("word")
        let date = Expression<Int>("date")
        let number = Expression<Int>("number")
        let address = Expression<String>("address")
        let notes = Expression<String>("notes")
        let units = Expression<String>("units")
        let responses = Expression<String>("responses")
        let updated = Expression<Int>("updated")

        var jsonEncodedNotes = "[]"
        if let jsonData = try? JSONSerialization.data(withJSONObject: alarm.notes, options: []) {
            jsonEncodedNotes = String(data: jsonData, encoding: .utf8)!
        }

        var jsonEncodedUnits = "[]"
        if let jsonData = try? JSONSerialization.data(withJSONObject: alarm.units, options: []) {
            jsonEncodedUnits = String(data: jsonData, encoding: .utf8)!
        }

        var jsonEncodedResponses = [String: Any]()
        alarm.responses.forEach { key, value in
            jsonEncodedResponses[String(key)] = value.toJson()
        }
        var jsonEncodedResponsesString = "{}"
        if let jsonData = try? JSONSerialization.data(withJSONObject: jsonEncodedResponses, options: []) {
            jsonEncodedResponsesString = String(data: jsonData, encoding: .utf8)!
        }

        let result = try? database!.run(alarms.insert(
            id <- alarm.id,
            type <- alarm.type,
            word <- alarm.word,
            date <- Int(alarm.date.timeIntervalSince1970),
            number <- alarm.number,
            address <- alarm.address,
            notes <- jsonEncodedNotes,
            units <- jsonEncodedUnits,
            responses <- jsonEncodedResponsesString,
            updated <- Int(alarm.updated.timeIntervalSince1970)
        ))

        if result == nil {
            update(alarm: alarm)
        }
    }

    static func update(alarm: Alarm) {
        let alarms = Table("Alarm")
        let id = Expression<String>("id")
        let type = Expression<String>("type")
        let word = Expression<String>("word")
        let date = Expression<Int>("date")
        let number = Expression<Int>("number")
        let address = Expression<String>("address")
        let notes = Expression<String>("notes")
        let units = Expression<String>("units")
        let responses = Expression<String>("responses")
        let updated = Expression<Int>("updated")

        var jsonEncodedNotes = "[]"
        if let jsonData = try? JSONSerialization.data(withJSONObject: alarm.notes, options: []) {
            jsonEncodedNotes = String(data: jsonData, encoding: .utf8)!
        }

        var jsonEncodedUnits = "[]"
        if let jsonData = try? JSONSerialization.data(withJSONObject: alarm.units, options: []) {
            jsonEncodedUnits = String(data: jsonData, encoding: .utf8)!
        }

        var jsonEncodedResponses = [String: Any]()
        alarm.responses.forEach { key, value in
            jsonEncodedResponses[String(key)] = value.toJson()
        }
        var jsonEncodedResponsesString = "{}"
        if let jsonData = try? JSONSerialization.data(withJSONObject: jsonEncodedResponses, options: []) {
            jsonEncodedResponsesString = String(data: jsonData, encoding: .utf8)!
        }

        let result = try? database!.run(alarms.filter(id == alarm.id).update(
            type <- alarm.type,
            word <- alarm.word,
            date <- Int(alarm.date.timeIntervalSince1970),
            number <- alarm.number,
            address <- alarm.address,
            notes <- jsonEncodedNotes,
            units <- jsonEncodedUnits,
            responses <- jsonEncodedResponsesString,
            updated <- Int(alarm.updated.timeIntervalSince1970)
        ))
    }
}

class AlarmResponse {
    var note: String
    var time: Int
    var responses: [Int: AlarmResponseType]

    init(note: String, time: Int, responses: [Int: AlarmResponseType]) {
        self.note = note
        self.time = time
        self.responses = responses
    }

    static let jsonShorts = [
        "note": "n",
        "time": "t",
        "responses": "r",
    ]

    init(json: [String: Any]) {
        self.note = json[AlarmResponse.jsonShorts["note"]!] as! String
        self.time = json[AlarmResponse.jsonShorts["time"]!] as! Int
        self.responses = {
            var result = [Int: AlarmResponseType]()
            let decoded = json[AlarmResponse.jsonShorts["responses"]!] as! [String: Any]
            decoded.forEach { key, value in
                result[Int(key)!] = AlarmResponseType(rawValue: value as! Int)!
            }
            return result
        }()
    }

    func toJson() -> [String: Any] {
        return [
            AlarmResponse.jsonShorts["note"]!: note,
            AlarmResponse.jsonShorts["time"]!: time,
            AlarmResponse.jsonShorts["responses"]!: {
                var result = [String: Any]()
                responses.forEach { key, value in
                    result[String(key)] = value.rawValue
                }
                return result
            }()
        ]
    }

    enum AlarmResponseType: Int {
        case onStation = 0
        case under5 = 1
        case under10 = 2
        case under15 = 3
        case onCall = 4
        case notReady = 5
        case notSet = 6
    }
}
